File: src/my_ude_py/scripts/turn_camera_client.py
#!/usr/bin/env python3
import cv2
import rclpy
import random
from rclpy.node import Node
from ude_interface.srv import TurnCamera
from cv_bridge import CvBridge 
from functools import partial

TIME_DEFAULT_ = 5.0

class TurnCameraClientNode(Node): 
    def __init__(self):
        super().__init__("turn_camera_client") 
        
        self.declare_parameter("client_time", TIME_DEFAULT_)
        self.client_time_= self.get_parameter("client_time").value
        self.timer = self.create_timer(self.client_time_, self.call_change_angle)
        self.get_logger().info("Turn Camera Client Node has been started")
        

    def call_change_angle(self):
        x=random.randint(-100,100)/10
        self.call_turn_camera_server(x)


    def call_turn_camera_server(self, degree_turn):
        client = self.create_client(TurnCamera, "turn_camera")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for the server to be up... ")
        
        request = TurnCamera.Request()
        request.degree_turn = degree_turn
        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_turn_camera_resp, degree_turn=degree_turn))

    def callback_turn_camera_resp(self, future, degree_turn):
        try:
            response = future.result()
            self.get_logger().info("Degree turned is " + str(degree_turn))
            image = CvBridge().imgmsg_to_cv2(response.camera_image)
            cv2.imshow("Turn Camera Image", image)
            cv2.waitKey(0)
            cv2.destroyAllWindows()
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    node = TurnCameraClientNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_ude_cpp/src/turn_camera_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "ude_interface/srv/turn_camera.hpp"
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <cmath>

using std::placeholders::_1;

const double time_default_ = 5.0;

class TurnCameraClientNode : public rclcpp::Node 
{
public:
    TurnCameraClientNode() : Node("turn_camera_client")
    {
        this-> declare_parameter("client_time", time_default_);
        double client_time_ = this-> get_parameter("client_time").as_double();
            
        timer_=this->create_wall_timer(std::chrono::milliseconds((int) (1000*client_time_)), 
                                        std::bind(&TurnCameraClientNode::callChangedAngle,this));
        client = this->create_client<ude_interface::srv::TurnCamera>("turn_camera");
        RCLCPP_INFO(this->get_logger(), "Turn Camera Client Node has been started.");
        
    }

    

private:
    void callChangedAngle(){
        double x = (randomDouble() * 10.0)-5.0;
        threads_.push_back(std::thread(std::bind(&TurnCameraClientNode::callTurnCameraService, this, x)));
    }

    void callTurnCameraService(float degree_turn)
    { 
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server to be up...");
        }

        auto request = std::make_shared<ude_interface::srv::TurnCamera::Request>();
        request->degree_turn = degree_turn;

        client->async_send_request(request,std::bind(&TurnCameraClientNode::response_received_callback, this, _1));
    }

    void response_received_callback(rclcpp::Client<ude_interface::srv::TurnCamera>::SharedFutureWithRequest future){
        try
        {
            auto response = future.get();
            auto cv_ptr = cv_bridge::toCvCopy(response.second.get()->camera_image, "bgr8");
            auto image = cv_ptr->image;
            RCLCPP_INFO(this->get_logger(), "Degree turned is %lf", response.first->degree_turn);
            cv::imshow("Robot Camera Image", image);
            cv::waitKey(0);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        }
}

    double randomDouble()
    {
        return double(std::rand()) / (double(RAND_MAX) + 1.0);
    }

    rclcpp::Client<ude_interface::srv::TurnCamera>::SharedPtr client;
    rclcpp::TimerBase::SharedPtr timer_;
    std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurnCameraClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}